go: enable multi-error validation (#47)

akavel opened the pull request:
In generated Go code, allow returning a list of validation errors instead of only the first failed.

The implementation approach is based on the discussion in #47.

akonradi commented on tests/harness/go/main/harness.go:
```diff
@@ -27,28 +27,45 @@ func main() {
+
+func mergeReasons(reasons []string, err error) []string {
+	multi, ok := err.(interface{ AllErrors() []error })
+	if !ok {
```
Shouldn't this be some kind of fatal error? If the provided multierr doesn't implement the `AllErrors()` method, something is broken.

akavel replied:
Good point, thanks; I'll modify this.

akonradi commented on tests/harness/harness.proto:
```diff
@@ -14,4 +14,5 @@ message TestResult {
     bool   Error  = 2;
     string Reason = 3;
     bool   AllowFailure = 4;
+    repeated string AllReasons = 5;
```
Once we add support for multiple errors in all languages, I think we'll want to get rid of the duplication between Reason and AllReason. What do you think of making Reason repeated ([which is wire-compatible with singular](https://stackoverflow.com/questions/45849903/can-a-proto3-optional-field-be-changed-to-repeated-without-breaking-wire-compati)) and adding a `bool CheckMultipleErrors` that is set by the Go (and later C++, Java, Python) binaries that we can drop once it is set for all languages?

akonradi replied:
Thinking about this more, I'm now less sure this is the right approach. See PR comments.

akavel replied:
As to `repeated`, I didn't realize that, thanks! However, if we go that way, I don't currently see a way how the executor could detect if the language's executable does or does not support "lazy mode" based just on that. Just treating `len(Reasons)==1` as "does not support lazy mode" seems wrong, given that many cases do actually expect 1 failure. Would you prefer to add an extra field e.g. ".SupportsMultiError" in TestResult? Or do you have some other idea/preference? Hm, or would you prefer if the language's executable took the `.Failures` field from parsed TestCase, and compared the count on their side? (I assume that would lead to removing the comparison from the executor.) If I'm misunderstanding, I'd be super grateful for further guidance!

As to doing an extra check on the first error, I'll add that in the Go language's executable. I don't think I'm doing this now in the executor, so I may well be misunderstanding something in your suggestion; I think it might be easier for us to talk about this further once I submit the new commit; if it's gonna be not what you meant, please tell me, I'm perfectly ok with pulling commits back or amending them or whatever.

akonradi replied:
Sorry for not being clear. Yeah, your `.SupportsMultiError` field on TestResult is what I was calling `CheckMultipleErrors`. We need a way for the test harness to know that for some languages, it should expect the executable to return all errors, and for others just the one. Options are

1. hard-code the expectation per-language in the harness
2. include that information in the TestResult

Hard-coding per-language mean checking the `TestCase.Failures` field and then modifying it based on the language (i.e. for C++ if `.Failures = 5`, treat it as 1). Including the information in the TestResult means we can upgrade each language's support independently without touching the harness again, and without hard-coding. As we upgrade each language we make the test executable also set the field, and rely on the default value (unset by un-upgraded language executables) being false. Once all languages support multiple errors, we can remove the field entirely and get rid of the false path.

akavel replied:
I guess I'll have to modify the other languages anyway, even if the modification is wire-compatible, given that the code for writing a `TestResult` in tests is auto-generated in CI anyway for all languages, so the modification of the `.Result[s]` field in `.proto` will spill and break the tests, I presume

akonradi replied:
Good point, I totally forgot about that. Thanks!

akonradi commented:
The implementation looks good, thanks! I think we should figure out what we want the end state of testing to look like so we can plan how to get there. In the long run, all PGV languages support lazy and full validation. Returning all errors from the harness executable seems like the right approach both now and in the future, and having the executor check that the full error count matches the expected count seems like the right way to do that. I'm torn between having the harness executable check on its own that the first error returned by lazy validation matches the one returned by full validation (which we'd have to implement in each harness executable, but isn't that complicated) versus making it simpler and doing the check in the executor as you've done here.

akavel commented:
@akonradi PTAL; I tried to implement what you suggested in the comments, to the extent I could understand it. I'd be super grateful if you could let me know how you see this, and if there are any further changes you'd like to see, also including modifying the last commits (I'm not 100% confident I correctly understood what you intended to communicate to me, I'd be happy to follow up on that & tweak!)

akavel commented:
[only rebased for now; didn't notice the reply in time, will proceed to implement it now]

akavel commented:
@akonradi PTAL ❤️ I did what I got from the above:
 - changed `Reason` to `repeated Reasons`,
 - did necessary tweaks in all languages to make them compile with the proto modified as above,
 - and added `CheckMultipleErrors`.
 
If you'd like something changed, or some cleanups in commits history, please do not be afraid to let me know! 😄

akonradi commented:
Thanks for pushing this through!

akavel commented:
@akonradi 🎉  wow also huge thanks, for persisting and finding time! ❤️

danielhochman commented:
This is really cool and will be very helpful in my application.

I would have preferred a non-breaking interface change that would also allow for future expansion of validation options without breaking the interface each time.

Something like:
`func Validate(opts ...ValidateOption)`

Maybe if someone expands validation options in the future and they're looking at this comment they can go in that direction.

danielhochman commented:
After going through some of the other effects of this breaking change, this should have been introduced as an entirely new method, either `ValidateAll()` or `ValidateWithOptions(opts ...ValidateOption)`, as opposed to my earlier suggestion of using variadic params to keep the function call compatible. The interface itself needed to remain compatible.

Many applications were relying on code that looks like:
```go
type validator interface {
  func Validate() error
}

func check(m proto.Message) error {
  if v, ok := m.(validator); ok {
    return m.Validate()
  }
  return nil
}
```

Note that PGV does not publish the validator interface so any application that wants to check if a message implements must maintain their own copy of the interface and upgrading PGV will not cause any compilation failure since the interface is not imported from PGV. Due to this change, the validation code above will silently stop validating all messages since the messages no longer match the interface.

akonradi commented:
@danielhochman do you think we need to revert this and re-think the implementation?

danielhochman commented:
Given the likelihood of silently skipping validation (I have seen it in three different places in my services so far), I think that is probably the best path forward here. Revert and immediately cut another release to minimize the number of users that take up 0.6.0. Up to you and the other maintainers of course.

akonradi commented:
Sounds good. We can break stuff intentionally but let's not do it quietly.

akavel commented:
@danielhochman if you're depending on a project that explicitly says it has an unstable, alpha-stage API, and *expects to break it*, what you ***should have done*** when depending on an interface from it, is ***use an "interface assertion"*** somewhere in your project, as in:

```go
var _ validator = (*MyProtoMessage)(nil)
```

If you followed this good Go practice, you'd have perfectly fine "loud warning" that the interface changed.

I'm deeply saddened that you prefer to shoot down a feature, rather than admit you haven't done what you should have done to protect yourself from such situation, or try to contribute constructively by improving the feature yourself. I'm [heavily afraid that the reversal above, if not undone, will be a death blow for this feature in any shape from perspective of external contribution](https://github.com/envoyproxy/protoc-gen-validate/issues/47#issuecomment-830982607), seeing how much time and effort it took both sides (me as a contributor and @akonradi as maintainer) to make it happen in this shape, which improved things even if it was imperfect (which is always true for software projects, even more so in alpha stage), yet how easy it was for it to be shot down based on some poorly substantiated complaints. 

As to `ValidateWith`, on a very shallow surface that sounds like a great idea in theory. But [have you put any thought into how you'd actually even start implementing it for this project? Personally, I did, and I'd find the added complexity staggering.](https://github.com/envoyproxy/protoc-gen-validate/issues/47#issuecomment-830982607) Are you really sure [you are gonna need it?](https://en.wikipedia.org/wiki/YAGNI)

pdecks commented:
Just another perspective here, that even if someone did program defensively before the change, because the `Validate()` method is at the core of using this plugin, as demonstrated by the documentation (original and now-restored), it should have been safeguarded by introducing new methods as @danielhochman suggests. While engineers should expect some volatility with an alpha project, it is probably in the project's best interest for adoption to reduce the severity of that volatility by preserving contracts for the projects central functionality.

danielhochman commented:
I'd appreciate if you wouldn't make directed personal attacks, suggesting that somehow I am doing something wrong in my projects instead of having a reasoned discussion about potential issues that users might encounter with the current implementation. We can find a way forward, which could include returning to the original implementation and additional documentation giving the user best practices to avoid issues if the interface changes, though as you'll see below there might still be issues. Everyone else here is operating in good faith and trying to work towards the best outcome for the users and contributors of this project. I'm not here to "shoot down" anything.

> Are you up to walk the walk and constructively contribute code to this feature?

If you want me to attempt an alternative implementation for consideration based on your original PR I'd be happy to do so. I already contributed a fix to another open source project https://github.com/grpc-ecosystem/go-grpc-middleware/pull/417 that supports your changes. I was waiting to see what your thoughts were based on new comments by myself and other users of the project on the issues with silent interface failures, to see if maybe you had more thoughts on the two method approach based on additional comments.

> what you **should have done** when depending on an interface from it, is **use an "interface assertion**" somewhere in your project

We depend on another module to validate our protos https://github.com/grpc-ecosystem/go-grpc-middleware/blob/master/validator/validator.go.

That library cannot assert its interface against my projects protos. On the other hand, its copy of the `validator` interface is private. That project could choose to expose that interface for users of the library like me to assert against, which feels a bit strange but maybe it is correct. Even so there is an issue.

Generated code from multiple modules can be used in a Go application. Which means that I can end up with a mix of the old and new validation interfaces depending on what that particular version of the protoc plugin was used for the module in question. In order for me to catch this, I would not only need to assert against a proto message from my project, but for every single module I import from that has validate on its generated code. And I would need to continue to support every version of the interface that exists if protoc-gen-validate continues to modify the `Validate` signature and other libraries don't update their generated code.

Another thought I had was that maybe pgv should provide an exported interface as a source of truth for users to assert against rather than having others repeat it, could help in the case that the project plans to further change the interface over time, but it would still fail to address the multiple module mixed interface issue.

> **you should have written basic unit tests for the validation functionality**

A test exercising the validation code against one of my protos would catch this issue (and that's exactly how I caught it in my project), but there is again still the issue of mixed interfaces from multiple modules. Meaning that the unit test would need to include at least one generated proto message from every module that a user imports which has validation. Furthermore, a lot of users may be prone to miss test coverage for a third-party module because it is not common to report on third-party coverage within a project.

```go
grpc.NewServer(grpc.UnaryInterceptor(validator.UnaryServerInterceptor()))
```

This line is covered if I run the server init code but even with 100% reported coverage on project code it I didn't actually exercise the validation code in the third-party.

> the added complexity honestly staggering.

Here's how I would attempt this if you want me to take a pass at it (generated code example). You did a lot of excellent work in the original PR and have way more context so maybe I'm missing some of the complexity with this approach, but again I would be happy to work on it and see for myself if you are amenable to it.

```go
func Validate() error {
   return ValidateWithOptions(false)
}

func ValidateWithOptions(all bool) error {
  ...
}
```

akonradi commented:
First off, thanks to @danielhochman for pointing out that the path we chose made upgrading PGV dangerous. This sort of constructive criticism is a key part of any open-source project. What's also key is the assumption of good faith on the part of the actors, and insinuations to the contrary are neither helpful nor appreciated.

I chose to revert this PR because it introduced more risk to consuming projects than I feel comfortable with (see https://github.com/envoyproxy/protoc-gen-validate/issues/47#issuecomment-831426532). I followed software engineering best-practices by reverting to a known good state instead of trying to fix forward. As @pdecks noted, even though PGV doesn't promise an explicit API, it is in our best interest to not break our users. Now that we're again providing a stable version to consumers, I'm more than happy to continue the discussion on #47 and see this feature land for good.

htuch commented:
As senior maintainer on Envoy and envoyproxy organization, I'm going to just observe that rollback in this situation is totally legitimate. A rollback does not mean that the feature is permanently reverted or rejected and we're happy to work on introducing in a way that addresses the rollback concern. The alpha status of PGV actually is moot; many folks rely on PGV in production and we should be sensitive to that.

It's always good to remember to assume good faith and best intent, be excellent to one another.

akavel commented:
@danielhochman @akonradi @htuch  Sorry for getting emotional and overreacting; as I tried to convey (inbetween the cries), I'm very interested in finding a way forward indeed. Honestly, seeing the response here from all of you gives me some hope something may actually still happen here indeed; thanks to all of you for caring enough to respond! ❤️  Based on history of #47, I was *sincerely* worried it may silently die, and I didn't seem to read much constructive attitude from the discussion around the reversal PR; seeing what you wrote now in response to my outburst, I'm more than happy to write that off as a misunderstanding on my part! 

> If you want me to attempt an alternative implementation for consideration based on your original PR I'd be happy to do so. 

@danielhochman: **I'd be *really* grateful and love it if you'd try!** I can make an attempt to be available and provide some support/review if you'd like, or not do so if you prefer; I am honestly still not sure how to structure the code to make it reasonably sane with a two-method interface; if you have some thoughts and ideas around that&mdash;as you seem to do&mdash;I'd be excited to see some full-blown code exploration from you around that in a new PR. What's especially important for me, is to have *some* working version of this feature available and *merged* in a finite timeframe. Personally I'm ok with it not being perfect (if that matters in any way), as long as it's at least reasonable (some ideas thrown around in #47 resulted in e.g. ~exponential complexity). Notably, a project I'm using PGV in really needs this feature. Seeing the caring response from @akonradi and @htuch , I'm hopeful again that they'll be kind enough to continue being available to guide & merge this feature's design & implementation once you hopefully open a new PR!

>  I can end up with a mix of the old and new validation interfaces depending on what that particular version of the protoc plugin was used for the module in question.

Ah, now that's a good one; thanks for pointing this out, that's something I didn't think of indeed.

brandoncole commented:
I think the intent of this PR was great, and being able to do multi-error validation is very much something I'd like to see included as a feature in the nearish future.  The lack of that capability is what keeps a lot of validation logic today for me in Go which I would love to get rid of.  

@akavel I appreciate the attempt you made at getting this functionality in, even though it unfortunately encountered a couple unforeseen setbacks. @danielhochman I like the suggestion you made above for a possible way of making a small change that might help this functionality see the light of day.  

If there's anyway I can help or give encouragement to the folks collaborating here to see this through, please let me know.  I'd be happy to test or contribute if needed.

danielhochman commented:
Thanks for the feedback! I am going to go with the two method approach to keep things simple, and leave a note for the future that if we ever need to introduce a third method we should strongly consider variadic functional options.

Should have a PR ready to go by the end of this week, can hopefully get it merged and released the following week pending review.
